Source/DafnyCore/Rewriters/AllOpaqueRevealStmtInserter.cs

```diff
@@ -0,0 +1,531 @@
+using System;
+using System.Collections.Generic;
+using System.Diagnostics.Contracts;
+using System.Linq;
+using System.Linq.Expressions;
+using Microsoft.Boogie;
+using Microsoft.Dafny.Plugins;
+
+namespace Microsoft.Dafny; 
+
+public class AllOpaqueRevealStmtInserter : IRewriter {
+  public AllOpaqueRevealStmtInserter(ErrorReporter reporter) : base(reporter) { }
+
+  internal override void PostResolveIntermediate(ModuleDefinition moduleDefinition) {
+    Contract.Requires(moduleDefinition != null);
+
+    foreach (var decl in moduleDefinition.TopLevelDecls) {
+      if (decl is TopLevelDeclWithMembers cl) {
+        foreach (var member in cl.Members) {
+          if (member is ICallable and not ConstantField) {
+            var mem = (ICallable)member;
+            if (member is Function { ByMethodDecl: not null } f) {
+              mem = f.ByMethodDecl;
+            }
+
+            if (mem is Method { Body: not null } method) {
+              AddMethodReveals(method, Reporter);
+            } else if (mem is Function { Body: not null } func) {
+              AddFunctionReveals(func, Reporter);
+            }
+          } else if (member is ConstantField { Rhs: not null } cf) {
+            var subExpressions = cf.Rhs.SubExpressions;
+
+            foreach (var expression in subExpressions) {
+              if (expression is FunctionCallExpr funcExpr) {
+                var func = funcExpr.Function;
+
+                if (isRevealedAlongPath(func, moduleDefinition, new List<ModuleDefinition> { func.EnclosingClass.EnclosingModuleDefinition })) {
+                  if (!func.IsOpaque && func.DoesAllOpaqueMakeOpaque(Options)) {
+                    var expr = cf.Rhs;
+
+                    var revealStmt0 = BuildRevealStmt(new FunctionReveal(func, new List<NameSegment>()),
+                      expr.Tok, moduleDefinition);
+
+                    if (revealStmt0 is not null) {
+                      var newExpr = new StmtExpr(expr.Tok, revealStmt0, expr) {
+                        Type = expr.Type
+                      };
+                      cf.Rhs = newExpr;
+                    }
+                  }
+
+                  using IEnumerator<FunctionReveal> enumerator = GetEnumerator(func, func.EnclosingClass, new List<Expression> { cf.Rhs }, moduleDefinition);
+                  while (enumerator.MoveNext()) {
+                    var newFunc = enumerator.Current;
+
+                    var origExpr = cf.Rhs;
+                    var revealStmt = BuildRevealStmt(newFunc, cf.Rhs.Tok, moduleDefinition);
+
+                    if (revealStmt is not null) {
+                      var newExpr = new StmtExpr(cf.Rhs.Tok, revealStmt, origExpr) {
+                        Type = origExpr.Type
+                      };
+                      cf.Rhs = newExpr;
+                    }
+                  }
+                }
+              }
+            }
+          }
```

Comment:
There is a lot of nesting here (maximum in this scope is 24 spaces). Could you please find ways to decrease nesting level to at most 8 spaces typically, perhaps 12 maximum?

Reply:
Done